python/lsst/daf/butler/core/location.py

```diff
@@ -54,6 +55,81 @@ def path(self):
         """
         return os.path.join(self._datastoreRoot, self._uri.path.lstrip("/"))
 
+    @property
+    def component_path(self):
+        """Path corresponding to location of component of composite.
+
+        This is used in cases where a composite is broken into separate
+        files, combining the root file path with the component name.
+        When reading a file from the URI, it is possible that the file
+        will not exist if the composite was written as a single entity
+        and the URI fragment is referring to a location inside that entity.
+
+        This path includes the root of the `Datastore`. Returns the empty
+        string if no fragment has been specified as part of the URI.
+        """
+        if not self.fragment:
+            return ''
+
+        # We need to insert the fragment before the file extension because
+        # some I/O libraries insist on a particular extension.
+        filepath, extension = os.path.splitext(self.path)
+        return "{}#{}{}".format(filepath, self.fragment, extension)
```

Comment:
I didn't think the user was allowed to control where the datastore puts files. I did it this way because I want to group all the components together.